---
title: 'KCNI Summer Academy 2023: Day 4'
author: "Shreejoy Tripathy and Mel Davie"
date: "13/07/2023"
output: github_document
---

# Case-control differential expression with pseudobulks
In this tutorial, we will:
*pseudobulk our single-cell data
*look at gene expression between cases and controls ("dementia" vs. "normal")
*fit a linear model to the data and perform differential gene expression (DGE)
*visualize DGE results
*perform a gene ontology enrichment analysis

*What are pseudobulks, and why are we using them?* When bulk RNA-sequencing data is collected, we get many cells representing a number of different cell types all sequenced together in the same sample, and we can't know which cell types are responsible for the gene expression changes we see between individuals. In single-cell RNA-seq, we have the benefit of being able to determine the cell type for each sample, but we also a data structure that reflects a lot of within-cell type variability that can reduce the signal-to-noise ratio for between-individual comparisons. Since we're interested in making comparisons at the cell type-level between individuals, we "pseudobulk" our single-cell data to get counts that are similar to what we'd get if we'd been able to perform bulk RNA-sequencing on just one cell type at a time in each individual.

## Setup
### Set working directory
```{r setup}
knitr::opts_knit$set(root.dir = "~/Documents/KCNISS_2023/")
```

### Load packages
```{r}
# install.packages("BiocManager")
# BiocManager::install("edgeR")
# install.packages("ggrepel")
# install.packages("cowplot")
# install.packages("gprofiler2")

### for older versions of R, may also need to run this:
#BiocManager::install("locfit") 
### if locfit above doesn't work (because your version of R is too old):
#install.packages("https://cran.r-project.org/src/contrib/Archive/locfit/locfit_1.5-9.4.tar.gz", repos=NULL, type="source")

suppressPackageStartupMessages({
  library(tidyverse)
  library(Seurat)
  library(edgeR)
  library(ggrepel)
  library(cowplot)
  library(limma)
  library(ggpubr)
  library(gprofiler2)
})
```

### Pseudobulking of a Seurat count matrix by donors and cell types
We're going to first illustrate pseudobulking using the 10x SEA-AD dataset that we used in the prior days. 

Note that pseudobulking should be performed on the full dataset, so we have done this for you and provided the output that will be used for the remainder of this tutorial. However, to illustrate that process, we will have you pseudobulk the mini SEA-AD dataset.

#### Create Seurat object
Load in the mini SEA-AD dataset - we can also just load in our older Seu_10x dataset as well.

```{r}
### metadata
meta_10x <- read.csv("./SEA-AD_10x_meta_mini.csv", row.names = 1)
#row.names(meta_10x) should contain same values as row.names(counts_10x)

### counts matrix
counts_10x <- read.csv("./SEA-AD_10x_counts_mini.csv", row.names = 1)

### create Seurat object
Seu_10x <- CreateSeuratObject(counts = t(counts_10x), 
                                meta.data = meta_10x)
```

#### Create the pseudobulk dataset
To do this, we will use a Seurat function called **AggregateExpression**, which can be used to sum expression values by some specified identity class.

The metadata columns that we will pseudobulk on are named: external_donor_name_label, subclass_label. Thus, in our pseudobulked dataset, columns denote expression from each donor for each cell type.
```{r}
### Inspect the metadata to get a sense of what information we could use to pseudobulk
head(meta_10x) %>% knitr::kable()

### We want our counts matrix such that each column is one cell type from one individual
pseudobulk_seu_10x <- AggregateExpression(object = Seu_10x, 
                                          group.by = c('external_donor_name_label', 
                                          'subclass_label'), 
                                          slot = 'counts')

### Extract just the pseudobulked count matrix as a dataframe
pseudobulk_seu_10x_counts <- pseudobulk_seu_10x[['RNA']] %>% as.data.frame()
                 
### Inspect the resulting object - note how the column names are a combination of 
### external_donor_name_label and subclass_label
head(pseudobulk_seu_10x_counts) %>% knitr::kable()
```

### Using pre-generated pseudobulks from the full SEA-AD dataset for a specific cell type - Astrocytes
The data we're about to load in have been pseudobulked in advance using this script:

https://github.com/stripathy/pseudobulk_tutorial/blob/master/sea_ad_make_pseudobulk.R

For our the past couple of days and in our pseudobulking example, we were working with a smaller version of the SEA-AD 10x dataset that only included samples from 5 controls. Now, we are working with samples from all donors, but for the sake of time, we will only examine one cell type: **astrocytes**.

#### Load pseudobulked gene expression data
We will load in three different files: 
*the pseudobulked counts matrix (that we prepared for you),
*a dataframe containing gene IDs/symbols, and
*metadata containing information about each of the 88 donors in our dataset.

Note that the counts matrix and the gene IDs file have the same number of rows - order here is important, as each row in the counts matrix corresponds to a row in the gene IDs file.

**Question:** why not just have the gene IDs as row names for the counts matrix like you may have done previously - why have two separate files? You'll notice that other information about the genes is stored in the gene_info file, and we want to keep this separate from the counts matrix (we can't have both numeric and character vectors in a matrix object).
```{r}
pseudobulk_astro_counts <- read_csv('astro_pseudobulk/astro_pseudobulk.csv')
gene_info = read_csv('astro_pseudobulk/astro_pseudobulk_gene_info.csv')
sea_ad_meta = read_csv('astro_pseudobulk/astro_pseudobulk_meta_updated.csv')

#rownames(pseudobulk_astro_counts) <- gene_info$gene_symbol
```

#### Preparing for subsequent analyses
Our ultimate goal is to compare gene expression in astrocytes from cases to that of controls, so we want to set up the case/control variable (called "disease" in our metadata) such that cases form one group, controls form another, and our DGE results provide information on cases relative to controls. Let's take a look at the current structure of the *disease* variable and make changes to set up the following analyses. 
```{r}
### Examine the current structure of the disease variable
str(sea_ad_meta$disease) 
#we have a character vector, which does not acknowledge that any of the values are repeated (no group structure)

### Redefine the disease variable as a factor with 2 levels, where the 2nd level will be considered relative to the first
sea_ad_meta <- sea_ad_meta %>% mutate(disease = factor(disease, 
                                                      levels = c('normal', 'dementia')))

### Examine our new structure
str(sea_ad_meta$disease)
```

The full SEA-AD dataset as downloaded also includes previously collected data from what are referred to as "neurotypical donors," which could be used to define cell types. For the purpose of our analyses, we want to exclude these and look only at the core SEA-AD dataset.
```{r}
### Filter out the pseudobulks from neurotypical references
sea_ad_meta <- sea_ad_meta %>% 
  filter(`Neurotypical reference` == F)

### Keep only the counts corresponding to the donors remaining post-filter
pseudobulk_astro_counts <- pseudobulk_astro_counts[sea_ad_meta$donor_id]
```

## Examining our dataset
**Question:** how does the number of astrocytes per person compare between controls and patients with dementia?
```{r}
p1 <- ggplot(sea_ad_meta, aes(x = disease, y = cells_per_donor, fill = disease)) + 
  geom_boxplot() + 
  theme_cowplot() + 
  ylab('Astrocyte nuclei per donor (count)') +
  stat_compare_means() # this adds a p-value from an automatically selected significance test
                       # that compares means (you can change this using the method parameter)

p2 <- ggplot(sea_ad_meta, aes(x = disease, y = cells_per_donor / total_cells_per_donor * 100, 
                              fill = disease)) + 
  geom_boxplot() + 
  theme_cowplot() + 
  ylab('Astrocyte nuclei fraction (%, total nuclei)') +
  stat_compare_means()

p1 + p2
```

It looks like the total number of astrocytes is similar between individuals with dementia and controls, but they make up a greater proportion of all collected cells in individuals with dementia. 

**Follow-up question:** What could cause this result?

**Question:** is astrocyte nuclei fraction correlated with our AD pathology score? (Note that the AD pathology score we are using is called continuous pseudo-progression score or CPS, check out the SEA-AD paper for more info on how this was defined: https://doi.org/10.1101/2023.05.08.539485)
```{r, message=FALSE}
ggplot(sea_ad_meta,
         aes(x = `Continuous Pseudo-progression Score`, y = cells_per_donor / total_cells_per_donor * 100)) + 
  geom_point(aes(color = disease)) + 
  geom_smooth(method = 'lm', se = F, color = "black") + 
  theme_cowplot() + 
  ylab('Astrocyte nuclei fraction (%, total nuclei)') +
  xlab('AD Pathology score') + 
  stat_cor(show.legend = F) 
```

Following up on our previous result, it looks like the proportion of astrocytes increases with CPS.

### Normalize our counts to CPM
If we want to plot the expression of a specific gene directly, we should first perform some normalization to account for technical variation (like sequencing depth variation) between the samples collected from each donor. For this, we compute counts per million (CPM).
```{r}
### Create a cpm object with normalized gene expression counts for each pseudobulk sample
pseudobulk_astro_cpm <- edgeR::cpm(y = pseudobulk_astro_counts) %>% 
  t() %>% # transpose the matrix so that we have our donors as the rows and genes as the columns
  as.data.frame() # convert the matrix to a data.frame for ggplot

colnames(pseudobulk_astro_cpm) <- gene_info$feature_name # add gene names

pseudobulk_astro_cpm <- left_join(sea_ad_meta, pseudobulk_astro_cpm %>% 
                                       rownames_to_column(var = 'donor_id')) # add on the metadata
```

### Plot gene expression (GFAP below) in case vs. control
Note that we're plotting log2(count + 1), a common workaround for the log of zero issue
```{r}
ggplot(pseudobulk_astro_cpm, aes(x = disease, y = log2(GFAP+1), fill = disease)) + 
  geom_boxplot() + 
  theme_cowplot() + 
  ylab('GFAP mRNA expr (log2 CPM+1)') +
  stat_compare_means()
```

It seems that GFAP (a canonical astrocyte marker) is expressed similarly between controls and individuals with dementia. This makes sense, as you can recall from our earlier plot that the number of astrocyte nuclei per donor is also similar between cases and controls.

## Differential expression with limma-voom
Next, we want to systematically determine which genes are differentially expressed between cases and controls. When doing this, we want to ensure that we account for any changes in gene expression that are tied to other variables like sex, post-mortem interval (PMI; a proxy for tissue quality and, by extension, of RNA quality), age, total RNA quantity (i.e., library size), and the total number of cells in each pseudobulk.

There are many different tools for performing a differential gene expression analysis—we will be using a workflow called **limma-voom**.

### Set up statistical model
First, let's figure out which donors we can use for this analysis. We need to subset to only include donors for which our metadata contains information on all of the variables we intend to account for in our model.

#### Select subjects
This next part gets a bit into Boolean variables or logicals, which are useful when you want to check if a list of conditions are simultaneously met.
```{r}
### Find subjects with information for disease, Sex, PMI, and Age
use_subjects <- complete.cases(sea_ad_meta %>% 
                                 dplyr::select(disease, Sex, PMI, Age_norm))
# the complete.cases function returns a TRUE/FALSE for each row, so we get a vector with a length matching
# the number of rows in our metadata. TRUE is returned only when a row has no missing values in all selected columns

### the authors say there's a set of 11 subjects with really bad AD that messes up the mRNA in their cells
### and that they should be excluded from downstream analyses
severely_affected_subjects <- c('H20.33.028', 'H20.33.033', 'H20.33.037', 'H20.33.038',
       'H20.33.046', 'H21.33.009', 'H21.33.020', 'H21.33.033',
       'H21.33.034', 'H21.33.039', 'H21.33.045')

### Drop subjects with fewer than 50 astrocytes per donor, younger donors who are part of the neurotypical 
### reference, and subjects who are in the severly_affected_subjects list
use_subjects <- use_subjects & 
  (sea_ad_meta$cells_per_donor > 50) & 
  (!sea_ad_meta$`Cognitive status` == 'Reference') & 
  (!sea_ad_meta$donor_id %in% severely_affected_subjects)
# for the code above, note that every vector provided is the same length, and each is ordered in the same way
# to represent the same rows. Here, we're creating a final vector that will tell us for each row whether all
# specified conditions are met--if any one of them is unmet, the final value for that row will be FALSE

### Filter the metadata to only include donors with information for all desired variables
sea_ad_meta_sub <- sea_ad_meta[use_subjects, ]
```

#### Prepare variables for input to model
We next want to pull out the variables from our metadata that we intend to include in the model design that we will specify with model.matrix later on, ensuring that we keep only the values for our selected subjects.
```{r}
disease <- sea_ad_meta_sub %>% 
  pull(disease)
# recall that we previously set this variable as factored so that our results show
# gene expression changes in "disease" relative to "normal"

neuropath_score <- sea_ad_meta_sub %>% 
  pull(`Continuous Pseudo-progression Score`)

sex <- sea_ad_meta_sub %>% 
  pull(Sex)

pmi <- sea_ad_meta_sub %>% 
  pull(PMI)

age <- sea_ad_meta_sub %>% 
  pull(Age_norm)

cells_per_donor <- sea_ad_meta_sub %>% 
  pull(cells_per_donor)

library_size <- pseudobulk_astro_counts[, use_subjects] %>% 
  colSums

### Count the number of subjects for each disease group (are they relatively balanced?)
sea_ad_meta_sub %>% 
  group_by(disease) %>% 
  tally
```

### Set up DGEList object
In this section, we create an object containing our count matrix and associated gene info so that we can use this as input for edgeR and limma functions.

We then apply a gene filter that excludes genes that aren't detected in enough of our samples (because we don't want to analyze differential expression for any genes that aren't likely to be biologically relevant--we're trying to improve our signal-to-noise ratio).
```{r}
### Create DGEList object
dge = DGEList(pseudobulk_astro_counts[, use_subjects], genes = gene_info)

### Filter for genes that are detected in at least 80% of samples
## Step 1: Create a binary matrix indicating if gene is detected in each sample
gene_detection_matrix <- dge$counts >= 1 

## Step 2: Calculate the number of samples in which each gene is detected
gene_detection_counts <- rowSums(gene_detection_matrix)

## Step 3: Define the minimum number of samples required to express the gene
min_samples_expressing_gene <- sea_ad_meta_sub %>% 
  nrow * 0.8 
# The 0.8 here refers to the fraction of total samples that needs to express the gene
# This line just computes the number that represents 80% of the total number of samples

## Step 4: Create a logical vector to identify genes that are detected in enough samples
genes_meeting_threshold <- gene_detection_counts >= min_samples_expressing_gene

## Step 5: Subset the DGEList to retain only genes meeting the threshold
dge <- dge[genes_meeting_threshold, ]

# Note that you could do all of the above in one line like this:
# dge <- dge[rowSums(dge$counts >= 1) >= min_samples_expressing_gene, ]
```

With our filtered DGEList object, we then apply the **calcNormFactors** function from edgeR.
Similar to when we computed counts per million (CPM), this will provide us with normalized count values that are more comparable between samples. 

We typically compute CPM for the purpose of visualizations, but a function like calcNormFactors performs additional normalization steps that improve the reliability and sensitivity of our differential expression results. In addition to accounting for differences in sequencing depth like the CPM, it also by default uses a method called Trimmed Mean of M-values (TMM) that performs the following steps:
-Calculates the M-values for each gene, which represent the log-fold change in expression between each sample and a reference sample (by default, the sample whose count-per-million upper quartile is closest to the mean upper quartile).
-Removes a certain percentage of genes with the highest and lowest M-values. This trimming step helps to mitigate the influence of genes with extreme expression changes, which could be due to technical artifacts or biological outliers.
-Calculates the mean M-value of the remaining genes. This trimmed mean serves as a reference point for the dataset.
-Calculates a scaling factor for each sample by comparing its mean M-value to the trimmed mean M-value. This scaling factor is used to adjust the expression values for each sample so that they are more comparable and less affected by technical variability.
-Applies the calculated scaling factors to the gene expression values for each sample, effectively normalizing the data.

Overall, this method helps to mitigate technical artifacts, biological outliers, and compositional bias (differences in the relative abundance of transcripts/genes between samples due to factors like gene length or GC content).
```{r}
### Calculate scaling factors to convert raw counts into "effective counts"
dge <- calcNormFactors(dge, method = "TMM")
```

### Set up design based on the factors defined above
For modeling AD, we have two options: treat it as a binary variable, or use the continuous AD pathology score that has been provided. Let's try both!

#### 1. Binary
```{r}
design_bin <- model.matrix(~ age + pmi + sex + log10(cells_per_donor) + log10(library_size) + disease) 

### Perform voom transformation
vm_bin <- voom(dge, design_bin, plot = TRUE)

### Perform lmFit and eBayes
fit_bin <- lmFit(vm_bin, design_bin)
fit_bin <- eBayes(fit_bin)
```
Note the plot that **voom** outputs. We can see here that there's a steady decline in standard deviation (variation) as count size (mean expression) increases. If you go back and include the genes that *aren't* expressed in at least 80% of the samples, you would see a slightly different relationship, where for low mean expression values, variance increases with mean expression. This is called **overdispersion**, and it is essential that we try and correct for this before running voom to produce more stable variance estimates, thereby producing more reliable differential expression results.

**lmFit** is performing linear modeling, fitting a linear model to the voom-transformed data with the specified design matrix, allowing for the estimation of coefficients (gene expression differences) and their standard errors for each gene. 

**eBayes** is applying empirical Bayes shrinkage to these coefficient estimates, which stabilizes the variances and provides moderated t-statistics, improving the reliability of differential expression analysis results.

#### 2. Continuous
```{r}
design_con <- model.matrix(~ age + pmi + sex + log10(cells_per_donor) + log10(library_size) + neuropath_score) 
# neuropath score here is the continuous pseudoprogression score 

### Perform voom transformation
vm_con <- voom(dge, design_con, plot = TRUE)

### Perform lmFit and eBayes
fit_con <- lmFit(vm_con, design_con)
fit_con <- eBayes(fit_con)
```

### Analyze sex and disease terms
It's finally time to get our list of differentially expressed genes (DEGs) and visualize the results! 

We will look at DEGs between males and females as well as DEGs between cases and controls.

#### Analyze genes associated with sex
*Why look at sex—aren't we just interested in differences between cases and controls?* It is increasingly required in publications and grant applications to consider and report the potential influence of sex, as this has been recognized as a factor that can significantly affect health outcomes. 

To pull the DGE results, we use a function called **topTable**. Here, we will specify the coefficient we want our log fold changes to reflect (so, male vs. female), the number of genes we want to pull (Inf will grab all of them), and the multiple testing correction we want to apply to our p-values (Benjamini-Hochberg) to manage false positives.

```{r}
# What are our coefficients called again?
fit_con$cov.coefficients %>% knitr::kable()

### Binary
deg_table_sex_bin = topTable(fit_bin, coef = "sexMale",  n = Inf, sort = "none", 
                         adjust.method = "BH")

deg_table_sex_bin %>% 
  arrange(adj.P.Val) %>% 
  head() %>% 
  knitr::kable()

### Continuous
deg_table_sex_con = topTable(fit_con, coef = "sexMale",  n = Inf, sort = "none", 
                         adjust.method = "BH")

deg_table_sex_con %>% 
  arrange(adj.P.Val) %>% 
  head() %>% 
  knitr::kable()
```

Whether we used the binary or the continuous AD measure doesn't seem to have much influence here. You can notice that the top DEGs here are (unsurprisingly) related to the X and Y chromosomes.

#### Volcano plot of genes associated with sex
To get a general sense of the gene expression differences between the sexes, we can create what is called a **volcano plot** (because it looks like the mouth of a volcano). It plots -log10(p-value) against log2(fold change). We transform our p-values for this plot so that the points are more visible, and higher values now reflect a more significant result.

```{r}
deg_table_sex_con %>% 
  ggplot(aes(x = logFC, y = -log10(P.Value), label = feature_name)) + 
  geom_point() +
  geom_text_repel(data = subset(deg_table_sex_con, adj.P.Val < 0.05), 
                  aes(label = feature_name), 
                  vjust = 1.5) +
  geom_point(data = subset(deg_table_sex_con, adj.P.Val < 0.05), 
             aes(color = 'red')) + 
  theme_cowplot() +
  theme(legend.position = "none")
```

To the right of the volcano (positive logFC), we have genes that are up-regulated in males relative to females. To the left, we instead have genes that are up in females.

What separates the genes from the left and right hand sides of the plot? Does this make sense?

#### Analyze genes associated with disease
Next, let's pull out the DEGs for our disease variable.
```{r}
### Binary
deg_table_disease_bin = topTable(fit_bin, coef = "diseasedementia",  n = Inf, sort = "none", 
                             adjust.method = "BH")

deg_table_disease_bin %>% 
  arrange(adj.P.Val) %>% 
  head(20) %>% 
  knitr::kable()

### Continuous
deg_table_disease_con = topTable(fit_con, coef = "neuropath_score",  n = Inf, sort = "none", 
                             adjust.method = "BH")

deg_table_disease_con %>% 
  arrange(adj.P.Val) %>% 
  head(20) %>% 
  knitr::kable()
```

Notice that treating disease as binary versus continuous completely changes our top genes!

#### Volcano plot of genes associated with disease
```{r}
### Binary
deg_table_disease_bin %>% 
  ggplot(aes(x = logFC, y = -log10(P.Value), label = feature_name)) + 
  geom_point() +
  geom_text_repel(data = subset(deg_table_disease_bin, adj.P.Val < 0.1), 
                  aes(label = feature_name), 
                  vjust = 1.5) + 
  geom_point(data = subset(deg_table_disease_bin, adj.P.Val < 0.1), 
             aes(color = 'red')) + 
  theme_cowplot() +
  theme(legend.position = "none")

### Continuous
deg_table_disease_con %>% 
  ggplot(aes(x = logFC, y = -log10(P.Value), label = feature_name)) + 
  geom_point() +
  geom_text_repel(data = subset(deg_table_disease_con, adj.P.Val < 0.1), 
                  aes(label = feature_name), 
                  vjust = 1.5, max.overlaps = 20) + 
  geom_point(data = subset(deg_table_disease_con, adj.P.Val < 0.1), 
             aes(color = 'red')) + 
  theme_cowplot() +
  theme(legend.position = "none")
```

Comparing this volcano plot to our earlier one, we can see there are far fewer DEGs in our binary AD condition than between the sexes. Why might this be the case?

Now plotting our continuous AD metric, we can see that there are MANY more DEGs. Why do you think that is?

#### Plot gene expression of a couple of DE genes below
Let's now plot some of the DE genes from above to highlight their expression changes between cases and controls.
##### Binary
```{r}
### Binary
col8a1_plot <- ggplot(pseudobulk_astro_cpm[use_subjects, ],
  aes(x = disease, y = log2(COL8A1+1), fill = disease)) + 
  ylab('COL8A1 mRNA expr (log2 CPM+1)') + 
  geom_boxplot() + 
  stat_compare_means() + 
  theme_cowplot() 

lrrc7_plot <- ggplot(pseudobulk_astro_cpm[use_subjects, ],
  aes(x = disease, y = log2(`LRRC7-AS1`+1), fill = disease)) + 
  ylab('LRRC7-AS1 mRNA expr (log2 CPM+1)') + 
  geom_boxplot() + 
  theme_cowplot() +
  stat_compare_means()

col8a1_plot + lrrc7_plot
```

##### Continuous
```{r, warning=FALSE, message=FALSE}
col8a1_plot <- ggplot(pseudobulk_astro_cpm[use_subjects, ],
  aes(x = `Continuous Pseudo-progression Score`, y = log2(COL8A1+1), color = disease, group = 1) ) + 
  ylab('COL8A1 mRNA expr (log2 CPM+1)') + 
  xlab('AD Pathology score') + 
  geom_point() + 
  geom_smooth(color = "black", method = 'lm', se = F) + 
  theme_cowplot() + 
  stat_cor(show.legend = F) 

slc28a2_plot <- ggplot(pseudobulk_astro_cpm[use_subjects, ],
  aes(x = `Continuous Pseudo-progression Score`, y = log2(SLC28A2+1), color = disease, group = 1)) + 
  ylab('SLC28A2 mRNA expr (log2 CPM+1)') + 
  xlab('AD Pathology score') + 
  geom_point() + 
  geom_smooth(color = "black", method = 'lm', se = F) + 
  theme_cowplot() +
  stat_cor(show.legend = F)

col8a1_plot + slc28a2_plot
```

## Gene ontology analysis
So, we have a few lists of DE genes—now what? One thing you can do after you know which genes are up- and down-regulated across conditions is to perform a gene ontology or GO analysis. This will allow you to see if there are any functions or pathways that are enriched among your list of genes. 

### Sex
#### Get gene lists
Let's grab the names of the significantly up-regulated and down-regulated genes from our earlier table. Note that we order these lists so that the genes that change the most are at the top.
```{r}
# Genes that have higher expression in males relative to females
male_up_ordered <- deg_table_sex_con %>% 
  filter(P.Value < 0.05 & logFC > 0) %>% 
  arrange(desc(logFC)) %>% 
  .$gene_symbol

# Genes that have lower expression in males relative to females
male_down_ordered <- deg_table_sex_con %>% 
  filter(P.Value < 0.05 & logFC < 0) %>% 
  arrange(logFC) %>% 
  .$gene_symbol
```

#### Enrichment analysis
We can now supply our ordered gene lists to a function from the package *gprofiler2* called **gost**. Note that we indicate that our query is ordered—this will factor into the significance calculations for our enriched terms. 

There are a variety of sources for gene ontology terms you can select. We'll grab biological processes (GO:BP), cellular components (GO:CC), and molecular functions (GO:MF).

We can then grab the most significant terms from each of those sources within the up- and down-regulated queries separately. This is useful for selecting terms to highlight in visualizations!
```{r}
gp_sex <- gost(list("up-regulated" = male_up_ordered,
	  "down-regulated" = male_down_ordered), organism = "hsapiens",
		        ordered_query = TRUE, sources = c("GO:BP", "GO:CC", "GO:MF"))

# Top terms
gp_sex_terms <- gp_sex$result %>% 
  group_by(query, source) %>% 
  slice_min(order_by = p_value, n = 2) %>% 
  .$term_id
```

#### Plot results
There are many ways to highlight GO terms. Here's one that is built in with *gprofiler2* that can be used to highlight GO terms with high -log10(p-adj) values.

Don't focus too much on the code of this plot, this is just a demonstration of how much you can tweak these figures to improve the aesthetic.
```{r}
gostplot(gp_sex, capped = FALSE, interactive = FALSE, 
         pal = list(`GO:MF` = "#6DC8E8", `GO:CC` = "#FF9179", `GO:BP` = "#57BCAC")) +
  geom_text_repel(mapping = aes(label = ifelse(gp_sex$result$term_id %in% 
                                                 gp_sex_terms,
                                               term_name, "")), 
                  force = 40, max.overlaps = Inf, size = 3) +
  geom_point(size = 3, colour = ifelse(gp_sex$result$term_id %in% 
                                                 gp_sex_terms,
                                       "black", "white"),
             alpha = ifelse(gp_sex$result$term_id %in% 
                                                 gp_sex_terms,
                                       1, 0))
```

Can you make sense of any of these terms?

### AD Neuropathology - associated genes
#### Get gene lists
```{r}
# Genes that have higher expression in individuals with dementia relative to controls
disease_up_ordered <- deg_table_disease_con %>% 
  filter(adj.P.Val < 0.1 & logFC > 0) %>% 
  arrange(desc(logFC)) %>% 
  pull(gene_symbol)

# Genes that have lower expression in individuals with dementia relative to controls
disease_down_ordered <- deg_table_disease_con %>% 
  filter(adj.P.Val < 0.1 & logFC < 0) %>% 
  arrange(logFC) %>% 
  pull(gene_symbol)
```

#### Enrichment analysis
```{r}
gp_disease <- gost(list("up-regulated" = disease_up_ordered,
	  "down-regulated" = disease_down_ordered), organism = "hsapiens",
		        ordered_query = TRUE, sources = c("GO:BP", "GO:CC", "GO:MF"))

# Top terms
gp_disease_terms <- gp_disease$result %>% 
  group_by(query, source) %>% 
  slice_min(order_by = p_value, n = 2) %>% 
  pull(term_id)
```

#### Plot results
```{r}
gostplot(gp_disease, capped = FALSE, interactive = FALSE, 
         pal = list(`GO:MF` = "#6DC8E8", `GO:CC` = "#FF9179", `GO:BP` = "#57BCAC")) +
  geom_text_repel(mapping = aes(label = ifelse(gp_disease$result$term_id %in% 
                                                 gp_disease_terms,
                                               term_name, "")), 
                  force = 40, max.overlaps = Inf,size = 3) +
  geom_point(size = 3, colour = ifelse(gp_disease$result$term_id %in% 
                                                 gp_disease_terms,
                                       "black", "white"),
             alpha = ifelse(gp_disease$result$term_id %in% 
                                                 gp_disease_terms,
                                       1, 0))
```

Finally, do you see any terms that make sense as being affected by AD?
