---
title: 'KCNI Summer Academy 2023: Day 4'
author: "Mel Davie"
date: "13/07/2023"
output: pdf_document
---

# Case-control differential expression with pseudobulks
## Setup
### Set working directory
```{r setup}
knitr::opts_knit$set(root.dir = "~/KCNISS_2023/SST_pseudobulk/data")
```

### Load packages
```{r}
#install.packages("BiocManager")
#BiocManager::install("edgeR")
#BiocManager::install("locfit") 
##if locfit above doesn't work (because your version of R is too old):
#install.packages("https://cran.r-project.org/src/contrib/Archive/locfit/locfit_1.5-9.4.tar.gz", repos=NULL, type="source")
#install.packages("ggrepel")
#install.packages("cowplot")
#install.packages("gprofiler2")

suppressPackageStartupMessages({
  library(tidyverse)
  library(Seurat)
  library(edgeR)
  library(ggrepel)
  library(cowplot)
  library(limma)
  library(ggpubr)
  library(gprofiler2)
})
```

### Illustrate how to perform pseudobulking of a seurat cout matrix by donors and by cell types
We're going to first illustrate pseudobulking using the 10x SEA-AD dataset that we used in the prior days

Load in the mini-sea ad dataset - we can also just load in our older Seu_10x dataset as well
```{r}
# metadata
meta_10x <- read.csv("../data/SEA-AD_10x_meta_mini.csv", row.names = 1)
#row.names(meta_10x) should contain same values as row.names(counts_10x)

# counts matrix
counts_10x <- read.csv("../data/SEA-AD_10x_counts_mini.csv", row.names = 1)

# create Seurat object
Seu_10x <- CreateSeuratObject(counts = t(counts_10x), 
                                meta.data = meta_10x)
```

### now we're going to create the pseudobulk dataset where columns denote expression from each donor and cell type

We're going to first illustrate pseudobulking using the 10x SEA-AD dataset that we used in the prior days

```{r}
# the metadata columns that we will pseudobulk on are named: 
# external_donor_name_label, subclass_label

# PseuobulkExpression is a convenience function within Seurat for generating Pseudobulks
pseudobulk_seu_10x_object <- Seurat:::PseudobulkExpression(object = Seu_10x, group.by = c('external_donor_name_label', 'subclass_label'), pb.method = 'aggregate', slot = 'counts')

pseudobulk_seu_10x_counts = pseudobulk_seu_10x_object[['RNA']] %>% as.data.frame()
                 
# now let's inspect the resulting object - note how the column names are a combination of external_donor_name_label and subclass_label

pseudobulk_seu_10x_counts[1:5, 1:5]
```

### Using pre-generated pseudobulks from the full SEA-AD dataset for a specific cell type - SST cells

The data we're about to load in have been pseudobulked in advance using this script:
https://github.com/stripathy/pseudobulk_tutorial/blob/master/sea_ad_make_pseudobulk.R

We will only be working with one cell type: SST neurons. 


### Load pseudobulked gene expression data
The data we're about to load in have been pseudobulked in advance using this script:
https://github.com/stripathy/pseudobulk_tutorial/blob/master/sea_ad_make_pseudobulk.R

We will only be working with one cell type: SST neurons. 

```{r}
pseudobulk_sst_counts = read_csv('/external/rprshnas01/netdata_kcni/stlab/SEA-AD/SST_pseudobulk.csv')
# pseudobulk_sst_counts = read_csv('SST_pseudobulk.csv')
gene_info = read_csv('/external/rprshnas01/netdata_kcni/stlab/SEA-AD/SST_pseudobulk_gene_info.csv')
sea_ad_meta = read_csv('/external/rprshnas01/netdata_kcni/stlab/SEA-AD/SST_pseudobulk_meta_updated.csv')


# sea_ad_meta = read_csv('~/pseudobulk_tutorial/SST_pseudobulk_meta_updated.csv')

# sea_ad_meta = read_csv('SST_pseudobulk_meta.csv')
# gene_info = read_csv('SST_pseudobulk_gene_info.csv')
rownames(pseudobulk_sst_counts) = gene_info$gene_symbol

sea_ad_meta = sea_ad_meta %>% mutate(disease = factor(disease, 
                                                      levels = c('normal', 'dementia')))
```

## Examining our dataset
How does the number of SST cells per person compare between controls and patients with dementia?

```{r}
p1 = sea_ad_meta %>% 
  filter(`Neurotypical reference` == F) %>%
  ggplot(aes(x = disease, y = cells_per_donor, fill = disease)) + 
  #geom_point() + 
  geom_boxplot() + 
  theme_cowplot() + 
  ylab('SST nuclei per donor (count)') +
  stat_compare_means()

p2 = sea_ad_meta %>% 
  filter(`Neurotypical reference` == F) %>%
  ggplot(aes(x = disease, y = cells_per_donor / total_cells_per_donor * 100, fill = disease)) + 
  #geom_point() + 
  geom_boxplot() + 
  theme_cowplot() + 
  ylab('SST nuclei fraction (%, total nuclei)') +
  stat_compare_means()

p1 + p2
```

It looks like there are fewer SST cells in individuals with dementia than controls

```{r}
p1 = sea_ad_meta %>% 
  filter(`Neurotypical reference` == F) %>%
  ggplot(aes(x = `Continuous Pseudo-progression Score`, y = cells_per_donor / total_cells_per_donor * 100, color = disease, group = 1)) + 
  geom_point() + 
  geom_smooth(method = 'lm', se = F) + 
  #geom_boxplot() + 
  theme_cowplot() + 
  ylab('SST nuclei fraction (%, total nuclei)') +
  xlab('AD Pathology score') + 
  stat_cor()

p1
```
It looks like there are fewer SST cells in individuals with greater continuous AD neuropathology scores

### Normalize our counts to CPM
```{r}
### Create a cpm object with normalized gene expression counts for each pseudobulk sample
pseudobulk_sst_cpm = edgeR::cpm(y = pseudobulk_sst_counts)
rownames(pseudobulk_sst_cpm) = gene_info$feature_name

pseudobulk_sst_cpm_trans = pseudobulk_sst_cpm %>% 
  t() %>% 
  as.data.frame()

pseudobulk_sst_cpm_trans = left_join(sea_ad_meta, pseudobulk_sst_cpm_trans %>% 
                                       rownames_to_column(var = 'donor_id')) 
```

### Plot gene expression of a gene (SST below) vs. all cell types
```{r}
p1 = pseudobulk_sst_cpm_trans %>% 
  filter(`Neurotypical reference` == F) %>%
  ggplot(aes(x = disease, y = log2(SST+1), fill = disease)) + 
  geom_boxplot() + 
  theme_cowplot() + 
  ylab('SST mRNA expr (log2 CPM+1)') +
  stat_compare_means()

p1
```
It looks like the expression of the SST mRNA is lower in individuals with dementia than controls.

### Show how SST normalized expression is related to the number of cells sampled per donor
```{r}
p2 = pseudobulk_sst_cpm_trans %>% 
  filter(`Neurotypical reference` == F) %>%
  ggplot(aes(x = cells_per_donor, y = log2(SST+1), color = disease, group = 1)) + 
  geom_point() + 
  geom_smooth(method = "lm", se = F) +
  theme_cowplot() + 
  xlab('SST nuclei per donor') + 
  ylab('SST mRNA expr (log2 CPM+1)') +
  stat_cor()

p3 = pseudobulk_sst_cpm_trans %>% 
  filter(`Neurotypical reference` == F) %>%
  ggplot(aes(x = cells_per_donor / total_cells_per_donor * 100, y = log2(SST+1), color = disease, group = 1)) + 
  geom_point() + 
  geom_smooth(method = "lm", se = F) +
  theme_cowplot() + 
  xlab('SST nuclei fraction (%, total nuclei)') +
  ylab('SST mRNA expr (log2 CPM+1)') +
  stat_cor()

p2 + p3

```
But notice that this is mostly an artifact of there being fewer SST cells in people with dementia than controls. This illustrates that numbers of cells sampled per cell type is a really important factor to consider when performing differential expression analyses.

## Differential expression with limma-voom
### Set up stats model
```{r}
# Find subjects with information for disease, Sex, PMI, and Age
use_subjects = complete.cases(sea_ad_meta %>% dplyr::select(disease, Sex, PMI, Age_norm))

# the authors say there's a set of 11 subjects with really bad AD that messes up the mRNA in their cells
# and that they should be excluded from downstream analyses
severely_affected_subjects = c('H20.33.028', 'H20.33.033', 'H20.33.037', 'H20.33.038',
       'H20.33.046', 'H21.33.009', 'H21.33.020', 'H21.33.033',
       'H21.33.034', 'H21.33.039', 'H21.33.045')

# Drop subjects with fewer than 50 SST cells per donor and also younger donors who are part of the neurotypical reference and also subjects who are in the severly_affected_donors_list
use_subjects = use_subjects & (sea_ad_meta$cells_per_donor > 50) & (!sea_ad_meta$`Cognitive status` == 'Reference') & (!sea_ad_meta$donor_id %in% severely_affected_subjects)

use_subjects_donor_list = sea_ad_meta[use_subjects, 'donor_id'] %>% unlist

disease = factor(unlist(sea_ad_meta[use_subjects, 'disease']), 
                 levels = c('normal', 'dementia'))
neuropath_score = sea_ad_meta[use_subjects, "Continuous Pseudo-progression Score"] %>% unlist
sex = sea_ad_meta[use_subjects, 'Sex'] %>% unlist
pmi = sea_ad_meta[use_subjects, 'PMI'] %>% unlist
age = sea_ad_meta[use_subjects, 'Age_norm'] %>% unlist
cells_per_donor = sea_ad_meta[use_subjects, 'cells_per_donor'] %>% unlist
library_size = pseudobulk_sst_counts[, use_subjects] %>% colSums %>% unlist()

# Count the number of subjects for each disease group
sea_ad_meta[use_subjects, ] %>% 
  group_by(disease) %>% 
  tally
```

### Set up DGEList object
```{r}
dge0 = DGEList(pseudobulk_sst_counts[, use_subjects], genes = gene_info)

min_samples_expressing_gene <- sea_ad_meta[use_subjects, ] %>% nrow * 0.8 
# The 0.8 here refers to the fraction of total samples that needs to express the gene

dge0 = dge0[rowSums(dge0$counts >= 1) >= min_samples_expressing_gene, ] 
# This step filters genes such that they need to be detected in at least 80% of samples

# dge0 = DGEList(pb_counts, group = pb_metadata$Phenotype)
dge0 = calcNormFactors(dge0, method = "TMM")
```

### Set up design based on the factors defined above - but we're first going to model AD disease as a binary variable
```{r}
design = model.matrix(~ age + pmi + sex + log10(cells_per_donor) + log10(library_size) + disease) 
# design = model.matrix(~ age + pmi + sex + log10(cells_per_donor) + neuropath_score) # neuropath score here is the continuous pseudoprogression score 


# Perform voom transformation
vm = voom(dge0, design, plot = TRUE)

# Perform lmFit and eBayes
fit = lmFit(vm, design)
fit = eBayes(fit)
```
From the plot above, can see that genes with low counts have already been filtered out (otherwise we'd see a downward dip in the trend towards the left—if you see this, you need to go back and filter).

### Analyze sex and disease terms
#### Analyze genes associated with sex
```{r}
deg_table_sex = topTable(fit, coef = "sexMale",  n = Inf, sort = "none", 
                         adjust.method = "BH")

deg_table_sex %>% 
  arrange(adj.P.Val) %>% 
  head()
```

#### Volcano plot of genes associated with sex
```{r}
sex_volcano = deg_table_sex %>% 
  ggplot(aes(x = logFC, y = -log10(P.Value), label = feature_name)) + 
  geom_point() +
  geom_text_repel(data = subset(deg_table_sex, adj.P.Val < 0.05), 
                  aes(label = feature_name), 
                  vjust = 1.5) +
  geom_point(data = subset(deg_table_sex, adj.P.Val < 0.05), 
             aes(color = 'red')) + 
  theme_cowplot() +
  theme(legend.position = "none")

sex_volcano
```
Talk about what separates the genes from the left hand side of the plot from the right hand side of the plot. Does this make sense?

#### Analyze genes associated with disease
```{r}
deg_table_disease = topTable(fit, coef = "diseasedementia",  n = Inf, sort = "none", 
                             adjust.method = "BH")

deg_table_disease %>% 
  arrange(adj.P.Val) %>% 
  head(20)
```

#### Volcano plot of genes associated with disease
```{r}
disease_volcano = deg_table_disease %>% 
  ggplot(aes(x = logFC, y = -log10(P.Value), label = feature_name)) + 
  geom_point() +
  geom_text_repel(data = subset(deg_table_disease, adj.P.Val < 0.1), 
                  aes(label = feature_name), 
                  vjust = 1.5) + 
  geom_point(data = subset(deg_table_disease, adj.P.Val < 0.1), 
             aes(color = 'red')) + 
  theme_cowplot() +
  theme(legend.position = "none")

disease_volcano
```
Talk about how many fewer genes there are here differentially expressed with AD versus sex above. Why might this be the case?



#### Plot gene expression of a couple of DE genes below
Let's now plot some of the DE genes from above.
```{r}
ihih5_plot = pseudobulk_sst_cpm_trans %>% 
  filter(donor_id %in% use_subjects_donor_list) %>%
  ggplot(aes(x = disease, y = log2(ITIH5+1), fill = disease) ) + 
  ylab('ITIH5 mRNA expr (log2 CPM+1)') + 
  geom_boxplot() + 
  stat_compare_means() + 
  theme_cowplot() 

ddx25_plot = pseudobulk_sst_cpm_trans %>% 
  filter(donor_id %in% use_subjects_donor_list) %>%
  ggplot(aes(x = disease, y = log2(DDX25+1), fill = disease)) + 
  ylab('DDX25 mRNA expr (log2 CPM+1)') + 
  geom_boxplot() + 
  theme_cowplot() +
  stat_compare_means()

ihih5_plot + ddx25_plot

```
### Let's redo the analysis of differential expression above, but let's now model AD as a continuous variable based on neuropathology scores
```{r}

design = model.matrix(~ age + pmi + sex + log10(cells_per_donor) + log10(library_size) + neuropath_score) # neuropath score here is the continuous pseudoprogression score 


# Perform voom transformation
vm = voom(dge0, design, plot = TRUE)

# Perform lmFit and eBayes
fit = lmFit(vm, design)
fit = eBayes(fit)

deg_table_ad_neuropath = topTable(fit, coef = "neuropath_score",  n = Inf, sort = "none", 
                             adjust.method = "BH")

deg_table_ad_neuropath %>% 
  arrange(adj.P.Val) %>% 
  head(20)

```
#### Volcano plot of genes associated with AD continuous neuropathology
```{r}
neuropath_volcano = deg_table_ad_neuropath %>% 
  ggplot(aes(x = logFC, y = -log10(P.Value), label = feature_name)) + 
  geom_point() +
  geom_text_repel(data = subset(deg_table_ad_neuropath, adj.P.Val < 0.1), 
                  aes(label = feature_name), 
                  vjust = 1.5) + 
  geom_point(data = subset(deg_table_ad_neuropath, adj.P.Val < 0.1), 
             aes(color = 'red')) + 
  theme_cowplot() +
  theme(legend.position = "none")

neuropath_volcano
```
Notice how many more genes are differentially expressed with AD neuropathology compared to the binary normal / dementia AD variable above.

#### Plot gene expression of a couple of DE genes below
Let's now plot some of the DE genes from above.
```{r}
negr1_plot = pseudobulk_sst_cpm_trans %>% 
  filter(donor_id %in% use_subjects_donor_list) %>%
  ggplot(aes(x = `Continuous Pseudo-progression Score`, y = log2(NEGR1+1), color = disease, group = 1) ) + 
  ylab('NEGR1 mRNA expr (log2 CPM+1)') + 
  xlab('AD Pathology score') + 
  geom_point() + 
  geom_smooth(method = 'lm', se = F) + 
  theme_cowplot() + 
  stat_cor() 

prom1_plot = pseudobulk_sst_cpm_trans %>% 
  filter(donor_id %in% use_subjects_donor_list) %>%
  ggplot(aes(x = `Continuous Pseudo-progression Score`, y = log2(PROM1+1), color = disease, group = 1)) + 
  ylab('PROM1 mRNA expr (log2 CPM+1)') + 
  xlab('AD Pathology score') + 
  geom_point() + 
  geom_smooth(method = 'lm', se = F) + 
  theme_cowplot() +
  stat_cor()

negr1_plot + prom1_plot

```

## Gene ontology analysis
### AD Neuropathology - associated genes
#### Get gene lists
```{r}
# Genes that have higher expression in individuals with dementia relative to controls
disease_up_ordered <- deg_table_ad_neuropath %>% 
  filter(adj.P.Val < 0.1 & logFC > 0) %>% 
  arrange(desc(logFC)) %>% 
  pull(gene_symbol)

# Genes that have lower expression in individuals with dementia relative to controls
disease_down_ordered <- deg_table_ad_neuropath %>% 
  filter(adj.P.Val < 0.1 & logFC < 0) %>% 
  arrange(logFC) %>% 
  pull(gene_symbol)
```

#### Enrichment analysis
```{r}
gp_disease = gost(list("up-regulated" = disease_up_ordered,
	  "down-regulated" = disease_down_ordered), organism = "hsapiens",
		        ordered_query = TRUE, sources = c("GO:BP", "GO:CC", "GO:MF"))

# Top terms
gp_disease_terms <- gp_disease$result %>% 
  group_by(query, source) %>% 
  slice_min(order_by = p_value, n = 3) %>% 
  pull(term_id)
```

#### Plot results
```{r}
gostplot(gp_disease, capped = FALSE, interactive = FALSE, 
         pal = list(`GO:MF` = "#6DC8E8", `GO:CC` = "#FF9179", `GO:BP` = "#57BCAC")) +
  geom_text_repel(mapping = aes(label = ifelse(gp_disease$result$term_id %in% 
                                                 gp_disease_terms,
                                               term_name, "")), 
                  force = 40, max.overlaps = Inf,size = 3) +
  geom_point(size = 3, colour = ifelse(gp_disease$result$term_id %in% 
                                                 gp_disease_terms,
                                       "black", "white"),
             alpha = ifelse(gp_disease$result$term_id %in% 
                                                 gp_disease_terms,
                                       1, 0))
```

### Sex
#### Get gene lists
```{r}
# Genes that have higher expression in males relative to females
male_up_ordered <- deg_table_sex %>% 
  filter(P.Value < 0.05 & logFC > 0) %>% 
  arrange(desc(logFC)) %>% 
  .$gene_symbol

# Genes that have lower expression in males relative to females
male_down_ordered <- deg_table_sex %>% 
  filter(P.Value < 0.05 & logFC < 0) %>% 
  arrange(logFC) %>% 
  .$gene_symbol
```

#### Enrichment analysis
```{r}
gp_sex = gost(list("up-regulated" = male_up_ordered,
	  "down-regulated" = male_down_ordered), organism = "hsapiens",
		        ordered_query = TRUE, sources = c("GO:BP", "GO:CC", "GO:MF"))

# Top terms
gp_sex_terms <- gp_sex$result %>% 
  group_by(query, source) %>% 
  slice_min(order_by = p_value, n = 3) %>% 
  .$term_id
```

#### Plot results
```{r}
gostplot(gp_sex, capped = FALSE, interactive = FALSE, 
         pal = list(`GO:MF` = "#6DC8E8", `GO:CC` = "#FF9179", `GO:BP` = "#57BCAC")) +
  geom_text_repel(mapping = aes(label = ifelse(gp_sex$result$term_id %in% 
                                                 gp_sex_terms,
                                               term_name, "")), 
                  force = 40, max.overlaps = Inf, size = 3) +
  geom_point(size = 3, colour = ifelse(gp_sex$result$term_id %in% 
                                                 gp_sex_terms,
                                       "black", "white"),
             alpha = ifelse(gp_sex$result$term_id %in% 
                                                 gp_sex_terms,
                                       1, 0))
```
