---
title: 'KCNI Summer Academy 2023: Day 2'
author: "Mel Davie"
date: "11/07/2023"
output: github_document
---

# Intro to differential expression, cell type identification & visualizations
## Setup
### Set working directory
```{r setup}
knitr::opts_knit$set(root.dir = "~/Downloads/KCNISS_2023/")
```

### Load packages
```{r}
#install.packages("ggpubr")
#install.packages("ggbeeswarm")
library(tidyverse)
library(Seurat)
library(ggpubr)
library(ggbeeswarm)
```

### Load Seurat object
```{r}
Seu_smart <- readRDS("Seu_smart.rds")
```

## Finding cluster marker genes
### Set your active identity
This changes the grouping variable for your Seurat object for all of the following analyses. Because we are going to conduct comparisons between clusters, let's use the new column that FindClusters() added to our metadata called "seurat_clusters".

```{r}
Idents(Seu_smart) <- "seurat_clusters" #set active identity to our newly defined clusters
```

### Perform differential expression
*FindMarkers()* identifies genes that are differentially expressed between one cluster and all other clusters. The parameters ident.1, logfc.threshold, and min.pct specify the cluster to compare against, the minimum log-fold change threshold, and the minimum percentage of cells expressing the gene, respectively. This step helps identify genes that can potentially distinguish one cluster from other clusters, and can be used for cell type annotation.

We can use this same function to find genes that can be used to:
- Distinguish one cluster from another closely related cluster
- Distinguish one cluster from all other clusters present

*Note the defaults for this function:* logfc.threshold = 0.25, min.pct = 0.1

*Consider:* can also choose to set a value for min.diff.pct (which is -Inf by default). This allows you to only test genes with a minimum difference in the min.pct values between the compared groups.

```{r}
# get genes to distinguish cluster 8 from 3
cluster_8_v_3 <- FindMarkers(Seu_smart, ident.1 = 8, ident.2 = 3, 
                             logfc.threshold = log(2), min.pct = 0.5)

# genes to distinguish cluster 8, 2, or 13 from all other clusters
cluster_8_v_all <- FindMarkers(Seu_smart, ident.1 = 8, logfc.threshold = log(2), 
                               min.pct = 0.5)
cluster_2_v_all <- FindMarkers(Seu_smart, ident.1 = 2, logfc.threshold = log(2), 
                               min.pct = 0.5)
cluster_13_v_all <- FindMarkers(Seu_smart, ident.1 = 13, logfc.threshold = log(2), 
                                min.pct = 0.5)
```

*FindAllMarkers()* will find markers that distinguish each cluster from the others for all clusters at once. If you're looking to identify cell types for your clusters, this will be the one you want to run. Note that it takes a while.

```{r}
all_clusters <- FindAllMarkers(Seu_smart, logfc.threshold = log(2), min.pct = 0.50)
```

### Examine results
```{r}
# taking a look at most significantly DE marker genes
cluster_8_v_3 %>% head(n = 10) %>% knitr::kable()
cluster_8_v_all %>% head(n = 10) %>% knitr::kable()
cluster_2_v_all %>% head(n = 10) %>% knitr::kable()

# two different ways of looking for a specific gene in the output dataframes, 
# where genes are rownames
cluster_2_v_all %>% 
  filter(row.names(.)=="GRIN1")

cluster_2_v_all %>% 
  rownames_to_column("gene_name") %>% 
  filter(gene_name=="GRIN1")
```

## Using metadata for other comparisons
Now, assume that we've performed cell type annotation for all of our cells using the clusters we've defined and their marker genes, and these are now stored in "subclass_label" in our metadata.

Let's say we aren't just interested in comparing between cell types, but also comparing between cell types derived from male vs. female donors. Could also think to do something like this for a cell type-specific case-control comparison.

Let's focus in on cluster 8, which we would've identified as microglia. 
```{r}
# making new metadata column to see if any sex differences for the same cell type
Seu_smart@meta.data <- Seu_smart@meta.data %>% 
  mutate(sex_subclass = paste(donor_sex_label, subclass_label, sep="_"))

table(Seu_smart$sex_subclass) #see how many cells we have for each sex+subclass combo

Idents(Seu_smart) <- "sex_subclass" #setting this new column as our active identity

unique(Idents(Seu_smart)) #seeing what our options are for making comparisons

#finding genes that are DE in female-derived microglia vs male-derived
F_microglia_vs_M_microglia <- FindMarkers(Seu_smart, ident.1 = "F_Microglia", 
                                          ident.2 = "M_Microglia", 
                                          logfc.threshold = log(2), min.pct = 0.25)

F_microglia_vs_M_microglia %>% head(n = 10) %>% knitr::kable()

Seu_smart@assays$RNA@counts["GFAP",][1:10] %>% knitr::kable() #checking if a gene is present in matrix
#this is grabbing the number of reads per sample for the gene GFAP for samples 1-10
```

## Visualizations
We'll continue with cluster 8 (microglia) to generate a few different figures with Seurat's plotting functions.
```{r}
#getting top 6 marker genes for distinguishing cluster 8 cells, saving to plot below
features <- cluster_8_v_all %>% 
  head(n=6) %>% 
  row.names

Idents(Seu_smart) <- "seurat_clusters" #setting our active identity back to our clusters
```

### Violin Plot
*VlnPlot()* generates a violin plot to visualize the expression distribution of the selected marker genes (features) across different clusters. This plot provides insights into the expression patterns and variations of these genes within each cluster, including the specific expression levels and the presence of bimodal or skewed distributions.

```{r}
VlnPlot(Seu_smart, features = features)
```

### Feature Plot
*FeaturePlot()* creates a feature plot to visualize the expression of the selected marker genes (features) in a scatterplot-like format. This plot displays the expression levels of the genes on the x-axis and the fraction of cells expressing the genes on the y-axis, with each dot representing a cell. It helps visualize the cell-to-cell variation in gene expression and identify potential correlations or differences between clusters.

```{r}
FeaturePlot(Seu_smart, features = features)
```

### Dot Plot
*DotPlot()* generates a dot plot to visualize the expression of the selected marker genes (features) in a matrix-like format. Each row represents a gene, each column represents a cluster, and the size/color of the dots indicates the gene expression level. This plot helps identify genes that are specifically expressed or enriched in cluster 8 compared to other clusters.

```{r}
DotPlot(Seu_smart, features = features) + RotatedAxis()
```

### Heat Map
*DoHeatmap()* creates a heat map to visualize the expression patterns of the selected marker genes (features) across cells. The heat map represents the gene expression values as a color-coded matrix, with rows corresponding to genes and columns corresponding to cells. It provides a comprehensive view of gene expression similarities and differences across cells and clusters.

```{r}
DoHeatmap(subset(Seu_smart, downsample = 100), features = features, size = 3, 
          slot="data") +
  scale_fill_viridis_c()
```

## Using human 10x data
For more information about this dataset, check this page: https://portal.brain-map.org/atlases-and-data/rnaseq/human-mtg-10x_sea-ad

### Loading data
```{r}
# metadata
meta_10x <- read.csv("SEA-AD_10x_meta_mini.csv", row.names = 1)
#row.names(meta_10x) should contain same values as row.names(counts_10x)

# counts matrix
counts_10x <- read.csv("SEA-AD_10x_counts_mini.csv", row.names = 1)

# create Seurat object
Seu_10x <- CreateSeuratObject(counts = t(counts_10x), 
                                meta.data = meta_10x)

#rm(meta_10x, counts_10x)
```

### Pre-processing
#### Quality control (QC)
*Note:* no need with Allen Institute cells, but with other datasets, should first plot QC metrics (like nfeatures_RNA). If using a histogram, would see bimodal distribution, can designate one group as "bad quality".

**Common QC metrics to check:**
- The number of unique genes detected in each cell.
  - Low-quality cells or empty droplets will often have very few genes
  - Cell doublets or multiplets may exhibit an aberrantly high gene count
- Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)
- The percentage of reads that map to the mitochondrial genome
  - Low-quality / dying cells often exhibit extensive mitochondrial contamination
  - We calculate mitochondrial QC metrics with the PercentageFeatureSet() function, which calculates the percentage of counts originating from a set of features
  - We use the set of all genes starting with MT- as a set of mitochondrial genes

```{r, eval=FALSE}
# The [[ operator can add columns to object metadata. Can stash QC stats
Seu_10x[["percent.mt"]] <- PercentageFeatureSet(Seu_10x, pattern = "^MT-")

VlnPlot(Seu_10x, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# looks like mitochondrial genes were filtered out in advance

# FeatureScatter is typically used to visualize feature-feature relationships, can be used
# for anything calculated by the object, e.g. columns in object metadata, PC scores
FeatureScatter(Seu_10x, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

# If we wanted to use this info to subset our cells, could do so like this
#Seu_10x <- subset(Seu_10x, subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & 
#                    percent.mt < 5)
```

#### Normalization & clustering workflow
```{r}
Seu_10x <- Seu_10x %>%
    SCTransform(variable.features.n = 3000) %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:30) %>%
    RunUMAP(dims = 1:30) %>%
    FindClusters()

ElbowPlot(Seu_10x, ndims = 50)

p1 <- DimPlot(Seu_10x, reduction = "umap", group.by = "subclass_label", label=TRUE)
p2 <- DimPlot(Seu_10x, reduction = "umap", group.by = "seurat_clusters", label=TRUE)

p1 + p2
```

### Differential expression
Use Day 1 as a reference to pick some cell types of interest, find their markers, and visualize them.

#### Plotting marker genes
```{r}
all_markers <- FindAllMarkers(Seu_10x, logfc.threshold = log(2), min.pct = .2)
markers <- all_markers %>% 
  distinct(gene, .keep_all = T) %>% 
  group_by(cluster) %>% 
  slice_max(order_by = avg_log2FC, n = 3) %>% 
  pull(gene)

DoHeatmap(subset(Seu_smart, downsample = 100), features = markers, size = 3,
          slot="data") +
  scale_fill_viridis_c()
```

## 10x vs SMART-seq
Now, we've looked at two different human-derived single-cell RNA-sequencing datasets which use two different sequencing methodologies. Let's compare them.

Plus, we'll do a soft intro to ggplots!

### Number of genes detected
```{r}
# Store the nFeature_RNA data
features_10x <- Seu_10x$nFeature_RNA %>% data.frame(genes_detected = .,
                                                      method = "10x")
features_smart <- Seu_smart$nFeature_RNA %>% data.frame(genes_detected = .,
                                                      method = "smart")

# Combine the data for plotting
features_combined <- bind_rows(features_10x, features_smart)

# Create a plot to compare the number of detected genes between the two methods
# Box plot
ggplot(features_combined, aes(x = method, y = genes_detected, fill = method)) +
  geom_boxplot() +
  labs(x = "Method", y = "Number of detected genes") +
  ggtitle("Comparison of genes detected per sample between 10x and SMART-seq") +
  theme_classic() +
  stat_compare_means() + 
  theme(legend.position = "none")

# Violin plot + box plot
ggplot(features_combined, aes(x = method, y = genes_detected, fill = method)) +
  geom_violin() +
  geom_boxplot(width = 0.1, fill = "white", outlier.shape = NA) +
  labs(x = "Method", y = "Number of detected genes") +
  ggtitle("Comparison of genes detected per sample between 10x and SMART-seq") +
  theme_classic() +
  stat_compare_means() + 
  theme(legend.position = "none")
  

# Beeswarm plot + box plot
ggplot(features_combined, aes(x = method, y = genes_detected, color = method)) +
  geom_quasirandom(alpha = 0.6, width = 0.2) +
  geom_boxplot(width = 0.1, fill = "white", outlier.shape = NA) +
  labs(x = "Method", y = "Number of detected genes") +
  ggtitle("Comparison of genes detected per sample between 10x and SMART-seq") +
  theme_classic() +
  stat_compare_means() + 
  theme(legend.position = "none")
```

### Number of RNA molecules
```{r}
# Store the nCount_RNA data
RNA_counts_10x <- Seu_10x$nCount_RNA %>% data.frame(RNA_count = .,
                                                      method = "10x")
RNA_counts_smart <- Seu_smart$nCount_RNA %>% data.frame(RNA_count = .,
                                                      method = "smart")

# Combine the data for plotting
RNA_counts_combined <- bind_rows(RNA_counts_10x, RNA_counts_smart)

# Box plot
ggplot(RNA_counts_combined, aes(x = method, y = RNA_count, fill = method)) +
  geom_boxplot() +
  labs(x = "Method", y = "Number of RNA molecules") +
  ggtitle("Comparison of RNA molecules per sample between 10x and SMART-seq") +
  theme_classic() +
  stat_compare_means() + 
  theme(legend.position = "none")
```

### Another way of doing this:
```{r}
features_10x <- Seu_10x@meta.data %>% 
  select(nCount_RNA, nFeature_RNA, subclass_label) %>% 
  mutate(method = "10x")

features_smart <- Seu_smart@meta.data %>% 
  select(nCount_RNA, nFeature_RNA, subclass_label) %>% 
  mutate(method = "smart")

features_combined = bind_rows(features_10x, features_smart)

ggplot(features_combined, aes(x = nCount_RNA, y = nFeature_RNA, colour = subclass_label)) +
  geom_point(alpha = 0.3) + 
  facet_wrap(~method, scales = "free_y") +
  geom_smooth(method = "lm") +
  theme_classic()
```

# Save your Seurat object for next time
```{r, eval=FALSE}
saveRDS(Seu_10x, "Seu_10x.rds")
```
